src/context/project-settings.tsx

```diff
@@ -0,0 +1,175 @@
+import { createContext, useContext, useEffect, useReducer } from "react";
+import { FormProps } from "@rjsf/core";
+import { ProjectSettingsTabRoutes } from "constants/routes";
+
+interface ProjectSettingsState
+  extends Record<ProjectSettingsTabRoutes, boolean> {}
+
+type SaveAction =
+  | { type: "saveTab"; tab: ProjectSettingsTabRoutes }
+  | { type: "unsaveTab"; tab: ProjectSettingsTabRoutes };
+
+interface SaveTabDispatchState {
+  saveTab: (tab: ProjectSettingsTabRoutes) => void;
+  unsaveTab: (tab: ProjectSettingsTabRoutes) => void;
+}
+
+const saveReducer = (
+  state: ProjectSettingsState,
+  action: SaveAction
+): ProjectSettingsState => {
+  switch (action.type) {
+    case "saveTab":
+      return state[action.tab] ? state : { ...state, [action.tab]: true };
+    case "unsaveTab":
+      return !state[action.tab] ? state : { ...state, [action.tab]: false };
+    default:
+      return state;
+  }
+};
+
+type FormAction = {
+  type: "updateForm";
+  tab: ProjectSettingsTabRoutes;
+  formData: FormProps<any>["formData"];
+};
+
+interface FormState
+  extends Record<ProjectSettingsTabRoutes, FormProps<any>["formData"]> {}
+
+interface FormContextState {
+  formState: FormState;
+  setFormState: (
+    tab: ProjectSettingsTabRoutes,
+    formData: FormProps<any>["formData"],
+    save?: boolean
+  ) => void;
+}
+
+const formReducer = (state: FormState, action: FormAction): FormState => {
+  switch (action.type) {
+    case "updateForm":
+      return {
+        ...state,
+        [action.tab]: action.formData,
+      };
+    default:
+      return state;
+  }
+};
+
+const ProjectSettingsContext = createContext<ProjectSettingsState | null>(null);
+const SaveTabDispatchContext = createContext<SaveTabDispatchState | null>(null);
+const FormContext = createContext<FormContextState | null>(null);
```

Comment:
I think it would make sense to combine these 3 contexts into 1 context. That manages the state of all interactions on the page. 
1. Keeping track of initial data. 
2. Keeping track of changes in form data. 
3. And using that to manage the save state of each tab. 

The  state changes that are happening could be handled by one centralized state change instead of 
happening in different places and would reduce the amount of side effects that occur as the result of them. https://github.com/evergreen-ci/spruce/pull/907/files#diff-7baaa44e7e71b372734d844c07ce31e5ecf2f902baf3da6c38be5ec3205a5fa2R85-R89

A state transition could handle all the related side effects in one action. As opposed to making those in separate updates.

Something like
```
    case "updateForm": {
          if(state.originalFormData !== formData) {
             return {
                ...state,
                [action.tab]: action.formData, // represents the form data for the tab
                hasChanges: true
              };
          }else{
             return {
             ...state,
             hasChanges: false
             }
       }
   }
```
How do you feel about that?